### PR TITLE
MC-1410 Make the `-sys` crates actually use system installed SDK headers

### PR DESCRIPTION
**Pull request by jcape**
SGX has been shipping pkg-config (`.pc`) files as part of it's SDK for some time now, and is populating `PKG_CONFIG_PATH` via the `/opt/intel/sgxsdk/environment` script as well. This is the time-tested, canonical way to discovery how to build stuff against open-source libraries in the GNU/Linux ecosystem.

At the same time, the `-sys`-suffixed crates in the rust ecosystem (e.g. `clang-sys`, `openssl-sys`) work by compiling against the system-installed versions of those libraries (hence the `-sys`). In order to get our `-sys` crates off the ground, I'd originally copied the header files from the SDK installation dir into our source tree, but we should really follow the rust ecosystem conventions, which is primarily what this PR is about.

Since this PR touches code that decides how to get linker paths and libs, I took the opportunity to add in linkage against the "LVI-mitigated" static libraries provided by Intel, which will need to be part of the evolving bolt story. Since this also touched an error enum, I took the liberty of removing its dependency on the now-deprecated `failure` as well.

This PR:

 - Links enclaves with the "LVI-mitigated" static libs provided by Intel.
 - Removes the use of the `SGX_SDK` compile-time environment variable in favor of `pkg-config` detection.
 - Removes the copied include files.
 - Sets a hard dependency on SGX SDK 2.9.1 (really: 2.9.101.2), enforced via pkg-config.
 - Reduces the scope of `SgxEnvironment` to just the HW and IAS mode variables.
 - Adds a new `SgxLibraryCollection` trait (implemented for slices of [pkg_config::Library](https://docs.rs/pkg-config/0.3.17/pkg_config/struct.Library.html)) that uses a collection of detected libraries to build include paths, linker lists, and handle `rerun_if_changed` stuff, etc.
 - Updates `Edger8r` and `SgxSign` to detect their paths from the "libsgx_launch" package (this is still a bit hacky, hopefully Intel will improve this as the SGX ecosystem becomes more integrated with traditional Linux distros).
 - Removes `mc-util-build-sgx` use of `failure` in favor of `displaydoc`
 - Updates `mc-util-build-enclave` to use the pkg-config stuff.

### CODE REVIEW

File: consensus/enclave/build.rs [hunk 1]
```diff
@@ -2,15 +2,37 @@
...
-use cargo_emit::{rustc_link_lib, rustc_link_search};
+use cargo_emit::rustc_cfg;
 use mc_util_build_script::Environment;
-use mc_util_build_sgx::{Edger8r, SgxEnvironment, SgxMode};
+use mc_util_build_sgx::{Edger8r, SgxEnvironment, SgxLibraryCollection, SgxMode};
+use pkg_config::{Config, Error as PkgConfigError, Library};
+
+const SGX_LIBS: &[&str] = &["libsgx_urts", "libsgx_epid"];
+const SGX_SIMULATION_LIBS: &[&str] = &["libsgx_urts_sim", "libsgx_epid_sim"];
+const SGX_VERSION: &str = "2.9.101.2";
```
Comment (eranrund) [inline consensus/enclave/build.rs:12]: `// Changing this version is a breaking change, you must update the crate version if you do.` like in other places?

Comment (jcape) [reply consensus/enclave/build.rs:12]: Yeah, I'll update, though ideally this kind of stuff can be exported by the dependencies, similar to the other EDL crates, which should isolate the untrusted enclave from needing to deal with pkg-config detection directly.

File: sgx/epid-types-sys/build.rs [hunk 2]
```diff
@@ -51,9 +60,42 @@ impl ParseCallbacks for Callbacks {
...
+    cfg.exactly_version(SGX_VERSION)
+        .print_system_libs(true)
+        .cargo_metadata(false)
+        .env_metadata(true);
+    let libnames = if sgx.sgx_mode() == SgxMode::Simulation {
+        rustc_cfg!("feature=\"sgx-sim\"");
+        SGX_SIMULATION_LIBS
+    } else {
+        SGX_LIBS
+    };
```
Comment (sugargoat) [inline sgx/epid-types-sys/build.rs:75]: This is really clear. Overall, the build.rs's are really nice.

File: util/build/enclave/src/lib.rs [hunk 3]
```diff
@@ -590,16 +612,24 @@ impl Builder {
...
                 unsigned_enclave
                     .to_str()
                     .expect("Invalid UTF-8 in unsigned enclave path"),
             ])
             .args(&["-z", "relro", "-z", "now", "-z", "noexecstack"])
-            .args(&["--no-undefined", "-nostdlib"])
-            .arg(format!("-L{}", self.libdir.display()))
+            .args(&["--no-undefined", "-nostdlib"]);
+
+        // We're only using link paths here, actual linkage is hard-coded.
```
Comment (sugargoat) [inline util/build/enclave/src/lib.rs:627]: Can you explain a little more about this - where is the linkage hard-coded?

File: util/build/sgx/src/libraries.rs [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+    /// Retrieve all the unique libraries to link to in this collection.
+    fn libs(&self) -> HashSet<&str>;
+
+    /// Emit all the relevant cargo linkage instructions for this collection.
+    // This function can be deprecated once we're done with baidu.
+    fn emit_cargo(&self);
+}
+
+/// A blanket implementation fo the SGX libraries collection for a [`Library`] slice.
```
Comment (sugargoat) [inline util/build/sgx/src/libraries.rs:23]:
```suggestion
/// A blanket implementation of the SGX libraries collection for a [`Library`] slice.
```

File: util/build/sgx/src/libraries.rs [hunk 5]
```diff
@@ -0,0 +1,62 @@
+// Copyright (c) 2018-2020 MobileCoin Inc.
+
+use cargo_emit::{rustc_link_lib, rustc_link_search};
+use pkg_config::Library;
+use std::{collections::HashSet, path::Path};
+
+/// A trait which adds SGX functionality onto a [`Library`] collection.
```
Comment (sugargoat) [inline util/build/sgx/src/libraries.rs:7]: Cool.

File: util/build/enclave/src/lib.rs (hunk 3 above)
Comment (jcape) [reply util/build/enclave/src/lib.rs:627]: It's the linker command in `Builder::link_unsigned()`.

Comment (sugargoat) [reply util/build/enclave/src/lib.rs:627]: Ah I see - thank you :)

**Pull request merged**